### PR TITLE
Add support for @nullable trait in order to support MergePatch usecases

### PR DESCRIPTION
**Pull request by astridej**
This PR is an update of https://github.com/disneystreaming/smithy4s/pull/1151 to take into account the changes to default and required handling that have happened since and to also change the modelling a little in line with what was discussed in the Discord (in particular, modelling required vs nullable as independent of each other). The change was big enough that I ended up writing the code from scratch, but I definitely based it on the linked PR.

With the change, it is possible to mark a field as nullable in order to distinguish between absence and removal of a field. This is modelled in Scala via `Option[Nullable[A]]` - `Nullable` being isomorphic to `Option` and left as a separate class for clarity. Unfortunately, this opens up a can of worms regarding possible combinations with @required and @default, the vast majority of which make no sense but are still theoretically possible. I had to rework `Field.Modifier` to make sure all possible combinations were reflected and handled properly, and although I did my best to keep the resulting complexity from leaking out it does show up in some places 😬

I think I've now got this in roughly the state the original PR was in, so opening a draft PR in order to get feedback while I work on the missing pieces.

## PR Checklist (not all items are relevant to all PRs)
    
- [✅] Added unit-tests (for runtime code)
- [✅] Added bootstrapped code + smoke tests (when the rendering logic is modified)
- [👎] Added build-plugins integration tests (when reflection loading is required at codegen-time)
- [👎] Added alloy compliance tests (when simpleRestJson protocol behaviour is expanded/updated)
- [✅ ] Updated dynamic module to match generated-code behaviour
- [✅ ] Added documentation
- [✅ ] Updated changelog

### CODE REVIEW

File: modules/bootstrapped/src/generated/com/amazonaws/dynamodb/Endpoint.scala
Comment (astridej) [inline modules/bootstrapped/src/generated/com/amazonaws/dynamodb/Endpoint.scala]: I'm not exactly sure why the hint ordering changed on some of the generated code or whether that's a problem? If it is, I can dig into it.

File: modules/codegen/src/smithy4s/codegen/internals/IR.scala
Comment (astridej) [inline modules/codegen/src/smithy4s/codegen/internals/IR.scala:167]: This case class is unfortunately complex, so I tried to make sure most places using it didn't actually need to destructure it. The `typeMod` field is part of that, since a bunch of consumers of this class only care about whether the resulting type is wrapped in an `Option` (or `Nullable`, or `Option[Nullable[...]]`).

I did look into converting the underlying type representation to a `Nullable` before reaching the point where the modifier was generated - in which case in theory nothing here should need to change - but I was scared off by the fact that nothing in that part of the code seemed to be doing modifications of the underlying shape's type. I can have a second look if people think it's better there, though - I'm still not happy with how complex the modifier logic is now. 🤷‍♀️

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala
Comment (astridej) [inline modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:811]: I still need to test that this works the way it should. (Even if part of me wants to say that anyone who annotates a field with `@errorMessage @nullable @required @default` or the like is getting what they deserve 🙈 )

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala
Comment (astridej) [inline modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: See todo comment - given the extra complexity involving `Nullable`, the current set of `FieldTN` values doesn't seem to accurately reflect the state of the field. That said, everything _seems_ to be working fine and this part of the code wasn't touched further in the original PR, so I'm not sure if I need to update them.

File: modules/core/src/smithy4s/schema/Schema.scala
Comment (astridej) [inline modules/core/src/smithy4s/schema/Schema.scala]: See TODO comment...

This is necessary in order to get the new code in `DocumentDecoderSchemaVisitor` and `SchemaVisitorJCodec` (which I copied from the original PR) to work for `Nullable.schema`, but I'm not sure if it breaks anything else. The main alternatives I can see are to drop that check and just always delegate null values to the underlying decoder if a `nullable` hint is present (is that safe?), or define the new schema here instead of as a bijection from option.

File: modules/bootstrapped/src/generated/com/amazonaws/dynamodb/Endpoint.scala
Comment (Baccata) [reply modules/bootstrapped/src/generated/com/amazonaws/dynamodb/Endpoint.scala]: mmm it may depend on non-deterministic stuff (ordering of map and sets). Not a problem unless the CI starts shouting (which it might), in which case we'll force a ShapeId-based ordering or something.

File: modules/bootstrapped/src/generated/smithy4s/example/PatchableEdgeCases.scala
Comment (Baccata) [inline modules/bootstrapped/src/generated/smithy4s/example/PatchableEdgeCases.scala:12]: this is great, thank you !

File: modules/core/src/smithy4s/schema/Schema.scala
Comment (Baccata) [inline modules/core/src/smithy4s/schema/Schema.scala]: I don't think `.isOption` is used anymore. It was probably used transiently, but I think the change is fine. 

Probably worth adding a `case RefinementSchema` too, just in case.

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:811]: 😆 yeah unfortunately this project is very much subject to Murphy's law, so I'm thankful with the additional patterns

File: modules/core/src/smithy4s/schema/Schema.scala
Comment (astridej) [reply modules/core/src/smithy4s/schema/Schema.scala]: It shows up once in a place involving `flexibleCollectionSupport` in the JSON codec ( https://github.com/disneystreaming/smithy4s/blob/aaf8b1f034b7d3f97e6f250c97fa08c517686d28/modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala#L895 ) but I'm not sure what the logic there is doing.

And good point on the RefinementSchema - will add that!

File: modules/codegen/src/smithy4s/codegen/internals/IR.scala
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/IR.scala:167]: I'll give it a think, but worst case scenario I'm not too bothered : the internal representation is implementation detail, and even though the complexity does increase with the "nullable" requirement, it's still acceptable imho

File: modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala]: I think you're not gonna need to do anything else : this code is aimed at rendering smithy trait as smithy4s hints. Smithy traits are expressed using a "canonical" node representation, which doesn't take into consideration any other trait but the `@required` trait. 

This kinda implies that we should probably protect against the application of the `@nullable` trait on shapes that are traits (or used in traits). 

You can avoid worrying about it.

File: modules/core/src/smithy4s/Nullable.scala
Comment (Baccata) [inline modules/core/src/smithy4s/Nullable.scala]: I think it'd be good to match the signature of Option's fold : 

```
def fold[B](whenNull : => B)(whenValue: A => B) : B
```

File: modules/core/src/smithy4s/Nullable.scala
Comment (Baccata) [inline modules/core/src/smithy4s/Nullable.scala]:
```suggestion
  final case class Value[A](a: A) extends Nullable[A]
```

File: modules/codegen/src/smithy4s/codegen/internals/IR.scala
Comment (Baccata) [reply modules/codegen/src/smithy4s/codegen/internals/IR.scala:167]: I tried to think of alternative ways of modelling this but what potentially made the code look nicer (using composable binary flags, for instance) or custom extractors also lose us exhaustive pat-mat, so ... I'm fine with the seemingly ugliness if you can't find a better way of modelling it.

**Comment by Baccata** [comment]
@astridej, one important aspect of this work is gonna be the addition of tests to [this specification](https://github.com/disneystreaming/smithy4s/blob/aaf8b1f034b7d3f97e6f250c97fa08c517686d28/modules/bootstrapped/test/src/smithy4s/schema/FieldSpec.scala#L9). 

Most encoders now use the `foreachUnlessDefault` or `getUnlessDefault`, and we need to ensure that those behave in a somewhat coherent manner with respect to the `nullable` trait. 

I presume that this line may have to change in order to account for the `nullable` trait  

https://github.com/disneystreaming/smithy4s/blob/aaf8b1f034b7d3f97e6f250c97fa08c517686d28/modules/core/src/smithy4s/schema/DefaultValueSchemaVisitor.scala#L97

File: modules/core/src/smithy4s/Nullable.scala
Comment (astridej) [reply modules/core/src/smithy4s/Nullable.scala]: Shall do!

Also, just realising - would it make sense to put some type classes for `Nullable` in the `cats` module?

Comment (Baccata) [reply modules/core/src/smithy4s/Nullable.scala]: I suppose it would 👍

File: modules/codegen/src/smithy4s/codegen/internals/IR.scala
Comment (astridej) [reply modules/codegen/src/smithy4s/codegen/internals/IR.scala:167]: Yeah, that was pretty much my conclusion too 😓 . At least `typeMod` got rid of a bunch of the places that would have needed to destructure this case class otherwise.

**Comment by astridej** [comment]
> @astridej, one important aspect of this work is gonna be the addition of tests to [this specification](https://github.com/disneystreaming/smithy4s/blob/aaf8b1f034b7d3f97e6f250c97fa08c517686d28/modules/bootstrapped/test/src/smithy4s/schema/FieldSpec.scala#L9).
> 
> Most encoders now use the `foreachUnlessDefault` or `getUnlessDefault`, and we need to ensure that those behave in a somewhat coherent manner with respect to the `nullable` trait.
> 
> I presume that this line may have to change in order to account for the `nullable` trait
> 
> https://github.com/disneystreaming/smithy4s/blob/aaf8b1f034b7d3f97e6f250c97fa08c517686d28/modules/core/src/smithy4s/schema/DefaultValueSchemaVisitor.scala#L97

@Baccata  I just had a look at this and wrote some tests. I can push them, but I should note that it looks like with the current implementation, the code is actually totally independent of `nullable` and doesn't have to change at all 😄. Digging into it, it seems to come down to two things:

* in this part of the code, `nullable` is handled in a different place from required/default - we first transform the schema into a nullable schema which sets the nullable hint, and then we call `.optional`, `.required`, or `.field` depending on the required/default situation to create the `Field` from it. `getUnlessDefault` can then ignore the nullable trait in the same it can ignore any other type refinement or complex underlying schema for a field. The only thing that could go wrong is that it could screw up on calculating the default, but...
* although there's no special logic for handling `Document.DNull` in the nullable case here, it does the right thing anyway: on receiving null, the logic gets forwarded to `DefaultValueSchemaVisitor`, which attempts to calculate the default value for a schema for `Nullable[T]` and ends up with `Nullable.Null` via the bijection to `Option`, which is what we would've wanted anyway.

If I'm off-base here, let me know!

File: modules/codegen/src/smithy4s/codegen/internals/Renderer.scala
Comment (astridej) [reply modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:811]: 😂 OK, added another detailed smoke spec test that goes through the eight possible cases and makes sure it can retrieve the error message.

File: modules/core/src/smithy4s/schema/Field.scala
Comment (Baccata) [inline modules/core/src/smithy4s/schema/Field.scala:70]: actually the reason we didn't factor it was to avoid an un-necessary allocation to `Option`, so I'm gonna ask you to revert (feel free to add a comment to point that out)

File: modules/bootstrapped/test/src/smithy4s/schema/FieldSpec.scala
Comment (Baccata) [inline modules/bootstrapped/test/src/smithy4s/schema/FieldSpec.scala:55]: Thank you for the additional tests. I'd like to see one for the following additional cases : 

```scala
val requiredNullable = string.nullable.required[Foo]("a", _.a)
```

and

```scala
case class Bar(
  a: Option[Nullable[String]]
)  

val optionalNullable = string.nullable.optional[Bar]("a", _.a)
```

**Comment by Baccata** [comment]
> Digging into it, it seems to come down to two things: [...] 
> If I'm off-base here, let me know!

I think you may be right, but I'd love for the assumption to be validated with a couple additional [test cases](https://github.com/disneystreaming/smithy4s/pull/1408#discussion_r1497774667).

File: modules/core/src/smithy4s/schema/Field.scala
Comment (astridej) [reply modules/core/src/smithy4s/schema/Field.scala:70]: Done, thanks for letting me know!

File: modules/bootstrapped/test/src/smithy4s/schema/FieldSpec.scala
Comment (astridej) [reply modules/bootstrapped/test/src/smithy4s/schema/FieldSpec.scala:55]: Added! I also took the liberty of adding one for `string.optional` (with no default set), because we didn't have one before and I wanted to make sure that the behaviour I was getting on `string.nullable.optional` was the same as the one on `string.optional`.

Comment (Baccata) [reply modules/bootstrapped/test/src/smithy4s/schema/FieldSpec.scala:55]: Awesome, thank you very much !

**Comment by kubukoz** [comment]
I love the emojis in the checklist. Going to steal this for my own :P

File: modules/docs/markdown/04-codegen/01-customisation/13-nullable-values.md
Comment (Baccata) [inline modules/docs/markdown/04-codegen/01-customisation/13-nullable-values.md]:
```suggestion
The official smithy toolset does not offer anything to distinguish between an absence of value and a value set to null during (de) serialisation. 
```

**Review by Baccata — APPROVED** [review]
This is an amazing PR, both from the added value but also from the fact that it has not required much oversight, and also because it adds some pretty thorough tests. 

I am extremely thankful for this 🥇 

Before we merge, I'd appreciate if @kubukoz or @lewisjkl could offer a second pair of eyes for a review, considering it's a fairly big PR.

**Comment by astridej** [comment]
@Baccata thank you for the kind words! 😊 🥳 I'm glad the PR worked without major changes being needed myself - and that you appreciate my edge-case-driven testing approach :)

And thank you for your quick review and good guidance!

**Review by lewisjkl — APPROVED** [review]
This is great work! Thanks for including docs and updating existing docs too. Really appreciate all of it 🙏

**Pull request merged**
